File: java/src/main/java/yatzykata/yatzy/Yatzy.java
package yatzykata.yatzy;

import java.util.Arrays;
import java.util.Comparator;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import yatzykata.yatzy.utils.IntComparisonOperator;

public class Yatzy {

  private static final int SCORE_ALL_DICE_ARE_EQUAL = 50;
  private static final int SCORE_OF_ZERO = 0;
  private static final int ROLL_PLACED_ON_ONES = 1;
  private static final int ROLL_PLACED_ON_TWOS = 2;
  private static final int ROLL_PLACED_ON_THREES = 3;
  private static final int ROLL_PLACED_ON_FOURS = 4;
  private static final int ROLL_PLACED_ON_FIVES = 5;
  private static final int ROLL_PLACED_ON_SIXES = 6;
  private static final int DIE_MATCH_TWO_TIMES = 2;
  private static final int DIE_MATCH_THREE_TIMES = 3;
  private static final int DIE_MATCH_FOUR_TIMES = 4;
  private static final int LIMIT_TO_ONE_PAIR = 1;
  private static final int LIMIT_TO_TWO_PAIRS = 2;

  public static Integer chance(Integer... dice) {
    return Stream.of(dice).reduce(0, Integer::sum);
  }

  public static Integer yatzy(Integer... dice) {
    return areAllDiceEqual(dice) ? SCORE_ALL_DICE_ARE_EQUAL : SCORE_OF_ZERO;
  }

  private static boolean areAllDiceEqual(Integer... dice) {
    return Stream.of(dice).distinct().count() <= 1;
  }

  public static Integer ones(Integer... dice) {
    return getScoreDicePlacedOn(dice, ROLL_PLACED_ON_ONES);
  }

  public static Integer twos(Integer... dice) {
    return getScoreDicePlacedOn(dice, ROLL_PLACED_ON_TWOS);
  }

  public static Integer threes(Integer... dice) {
    return getScoreDicePlacedOn(dice, ROLL_PLACED_ON_THREES);
  }

  public static Integer fours(Integer... dice) {
    return getScoreDicePlacedOn(dice, ROLL_PLACED_ON_FOURS);
  }

  public static Integer fives(Integer... dice) {
    return getScoreDicePlacedOn(dice, ROLL_PLACED_ON_FIVES);
  }

  public static Integer sixes(Integer... dice) {
    return getScoreDicePlacedOn(dice, ROLL_PLACED_ON_SIXES);
  }

  public static Integer getScoreDicePlacedOn(Integer[] dice, Integer placedOn) {
    return Stream.of(dice).filter(die -> die.equals(placedOn)).reduce(0, Integer::sum);
  }

  public static Integer pair(Integer... dice) {
    Stream<Integer> diceFoundMultipleTimes =
        getDiceByExactlyANumberOfTimesADieIsFound(dice, DIE_MATCH_TWO_TIMES)
            .sorted(Comparator.reverseOrder())
            .limit(LIMIT_TO_ONE_PAIR);

    return getScoreForDiceThatMatchMultipleTimes(diceFoundMultipleTimes, DIE_MATCH_TWO_TIMES);
  }

  public static Integer twoPairs(Integer... dice) {
    Stream<Integer> diceFoundMultipleTimes =
        getDiceByAtLeastANumberOfTimesADieIsFound(dice, DIE_MATCH_TWO_TIMES)
            .limit(LIMIT_TO_TWO_PAIRS);

    return getScoreForDiceThatMatchMultipleTimes(diceFoundMultipleTimes, DIE_MATCH_TWO_TIMES);
  }

  public static Integer threeOfAKind(Integer... dice) {
    Stream<Integer> diceFoundMultipleTimes =
        getDiceByAtLeastANumberOfTimesADieIsFound(dice, DIE_MATCH_THREE_TIMES);

    return getScoreForDiceThatMatchMultipleTimes(diceFoundMultipleTimes, DIE_MATCH_THREE_TIMES);
  }

  public static Integer fourOfAKind(Integer... dice) {
    Stream<Integer> diceFoundMultipleTimes =
        getDiceByAtLeastANumberOfTimesADieIsFound(dice, DIE_MATCH_FOUR_TIMES);

    return getScoreForDiceThatMatchMultipleTimes(diceFoundMultipleTimes, DIE_MATCH_FOUR_TIMES);
  }

  private static Stream<Integer> getDiceByExactlyANumberOfTimesADieIsFound(
      Integer[] dice, Integer numberOfTimesDieIsFound) {
    return getDiceByNumberOfTimesDieIsFound(
        dice, numberOfTimesDieIsFound, IntComparisonOperator.EQUAL);
  }

  private static Stream<Integer> getDiceByAtLeastANumberOfTimesADieIsFound(
      Integer[] dice, Integer numberOfTimesDieIsFound) {
    return getDiceByNumberOfTimesDieIsFound(
        dice, numberOfTimesDieIsFound, IntComparisonOperator.GREATER_THAN_OR_EQUAL);
  }

  private static Stream<Integer> getDiceByNumberOfTimesDieIsFound(
      Integer[] dice,
      Integer numberOfTimesDieIsFound,
      IntComparisonOperator intComparisonOperator) {
    return getDiceAndNumberOfTimesEachDieIsFound(dice).entrySet().stream()
        .filter(
            dieAndNumberOfTimesFound ->
                intComparisonOperator.compare(
                    dieAndNumberOfTimesFound.getValue(), numberOfTimesDieIsFound))
        .map(Map.Entry::getKey);
  }

  private static Map<Integer, Integer> getDiceAndNumberOfTimesEachDieIsFound(Integer[] dice) {
    return Arrays.stream(dice)
        .collect(
            Collectors.groupingBy(
                Function.identity(),
                Collectors.collectingAndThen(Collectors.counting(), Long::intValue)));
  }

  private static Integer getScoreForDiceThatMatchMultipleTimes(
      Stream<Integer> diceFoundMultipleTimes, Integer numberOfTimesDieIsFound) {
    return diceFoundMultipleTimes.reduce(
        0, (score, die) -> score + (die * numberOfTimesDieIsFound));
  }

  public static int smallStraight(int d1, int d2, int d3, int d4, int d5) {
    int[] tallies;
    tallies = new int[6];
    tallies[d1 - 1] += 1;
    tallies[d2 - 1] += 1;
    tallies[d3 - 1] += 1;
    tallies[d4 - 1] += 1;
    tallies[d5 - 1] += 1;
    if (tallies[0] == 1 && tallies[1] == 1 && tallies[2] == 1 && tallies[3] == 1 && tallies[4] == 1)
      return 15;
    return 0;
  }

  public static int largeStraight(int d1, int d2, int d3, int d4, int d5) {
    int[] tallies;
    tallies = new int[6];
    tallies[d1 - 1] += 1;
    tallies[d2 - 1] += 1;
    tallies[d3 - 1] += 1;
    tallies[d4 - 1] += 1;
    tallies[d5 - 1] += 1;
    if (tallies[1] == 1 && tallies[2] == 1 && tallies[3] == 1 && tallies[4] == 1 && tallies[5] == 1)
      return 20;
    return 0;
  }

  public static int fullHouse(int d1, int d2, int d3, int d4, int d5) {
    int[] tallies;
    boolean _2 = false;
    int i;
    int _2_at = 0;
    boolean _3 = false;
    int _3_at = 0;

    tallies = new int[6];
    tallies[d1 - 1] += 1;
    tallies[d2 - 1] += 1;
    tallies[d3 - 1] += 1;
    tallies[d4 - 1] += 1;
    tallies[d5 - 1] += 1;

    for (i = 0; i != 6; i += 1)
      if (tallies[i] == 2) {
        _2 = true;
        _2_at = i + 1;
      }

    for (i = 0; i != 6; i += 1)
      if (tallies[i] == 3) {
        _3 = true;
        _3_at = i + 1;
      }

    if (_2 && _3) return _2_at * 2 + _3_at * 3;
    else return 0;
  }
}

File: go/yatzy/v1/yatzy.go
package v1

type YatzyStruct struct {
	dice []int
}

func NewYatzy(d1, d2, d3, d4, _5 int) YatzyStruct {
	return YatzyStruct{[]int{d1, d2, d3, d4, _5}}
}

func Chance(d1, d2, d3, d4, d5 int) int {
	total := 0
	total += d1
	total += d2
	total += d3
	total += d4
	total += d5
	return total
}

func Yatzy(dice []int) int {
	counts := make([]int, 6)
	for _, die := range dice {
		counts[die-1] += 1
	}

	for i := 0; i != 6; i++ {
		if counts[i] == 5 {
			return 50
		}
	}
	return 0
}

func Ones(d1, d2, d3, d4, d5 int) int {
	sum := 0
	if d1 == 1 {
		sum += 1
	}
	if d2 == 1 {
		sum += 1
	}
	if d3 == 1 {
		sum += 1
	}
	if d4 == 1 {
		sum += 1
	}
	if d5 == 1 {
		sum += 1
	}
	return sum
}

func Twos(d1, d2, d3, d4, d5 int) int {
	sum := 0
	if d1 == 2 {
		sum += 2
	}
	if d2 == 2 {
		sum += 2
	}
	if d3 == 2 {
		sum += 2
	}
	if d4 == 2 {
		sum += 2
	}
	if d5 == 2 {
		sum += 2
	}
	return sum
}

func Threes(d1, d2, d3, d4, d5 int) int {
	s := 0
	if d1 == 3 {
		s += 3
	}
	if d2 == 3 {
		s += 3
	}
	if d3 == 3 {
		s += 3
	}
	if d4 == 3 {
		s += 3
	}
	if d5 == 3 {
		s += 3
	}
	return s
}

func (y YatzyStruct) Fours() int {
	sum := 0
	for at := 0; at < 5; at++ {
		if y.dice[at] == 4 {
			sum += 4
		}
	}

	return sum
}

func (y YatzyStruct) Fives() int {
	s := 0
	i := 0
	for i = 0; i < len(y.dice); i++ {
		if y.dice[i] == 5 {
			s = s + 5
		}
	}

	return s
}

func (y YatzyStruct) Sixes() int {
	sum := 0
	for at := 0; at < len(y.dice); at++ {
		if y.dice[at] == 6 {
			sum = sum + 6
		}
	}

	return sum
}
func Score_pair(d1, d2, d3, d4, d5 int) int {
	counts := make([]int, 6)
	counts[d1-1] += 1
	counts[d2-1] += 1
	counts[d3-1] += 1
	counts[d4-1] += 1
	counts[d5-1] += 1
	at := 0
	for at = 0; at < 6; at++ {
		if counts[6-at-1] == 2 {
			return (6 - at) * 2
		}
	}
	return 0
}

func Two_pair(d1, d2, d3, d4, d5 int) int {
	counts := make([]int, 6)
	counts[d1-1] += 1
	counts[d2-1] += 1
	counts[d3-1] += 1
	counts[d4-1] += 1
	counts[d5-1] += 1
	n := 0
	score := 0
	for i := 0; i < 6; i++ {
		if counts[6-i-1] >= 2 {
			n = n + 1
			score += (6 - i)
		}
	}
	if n == 2 {
		return score * 2
	} else {
		return 0
	}
}

func Four_of_a_kind(_1, _2, d3, d4, d5 int) int {
	tallies := make([]int, 6)
	tallies[_1-1] += 1
	tallies[_2-1] += 1
	tallies[d3-1] += 1
	tallies[d4-1] += 1
	tallies[d5-1] += 1
	for i := 0; i < 6; i++ {
		if tallies[i] >= 4 {
			return (i + 1) * 4
		}
	}
	return 0
}

func Three_of_a_kind(d1, d2, d3, d4, d5 int) int {
	t := make([]int, 6)
	t[d1-1] += 1
	t[d2-1] += 1
	t[d3-1] += 1
	t[d4-1] += 1
	t[d5-1] += 1
	for i := 0; i < 6; i++ {
		if t[i] >= 3 {
			return (i + 1) * 3
		}
	}
	return 0
}

func SmallStraight(d1, d2, d3, d4, d5 int) int {
	tallies := make([]int, 6)
	tallies[d1-1] += 1
	tallies[d2-1] += 1
	tallies[d3-1] += 1
	tallies[d4-1] += 1
	tallies[d5-1] += 1
	if tallies[0] == 1 &&
		tallies[1] == 1 &&
		tallies[2] == 1 &&
		tallies[3] == 1 &&
		tallies[4] == 1 {
		return 15
	}
	return 0
}

func LargeStraight(d1, d2, d3, d4, d5 int) int {
	tallies := make([]int, 6)
	tallies[d1-1] += 1
	tallies[d2-1] += 1
	tallies[d3-1] += 1
	tallies[d4-1] += 1
	tallies[d5-1] += 1
	if tallies[1] == 1 &&
		tallies[2] == 1 &&
		tallies[3] == 1 &&
		tallies[4] == 1 &&
		tallies[5] == 1 {
		return 20
	}
	return 0

}

func FullHouse(d1, d2, d3, d4, d5 int) int {
	var tallies []int
	_2 := false
	i := 0
	_2_at := 0
	_3 := false
	_3_at := 0
	tallies = make([]int, 6)
	tallies[d1-1] += 1
	tallies[d2-1] += 1
	tallies[d3-1] += 1
	tallies[d4-1] += 1
	tallies[d5-1] += 1
	for i = 0; i < 6; i++ {
		if tallies[i] == 2 {
			_2 = true
			_2_at = i + 1
		}
	}
	for i = 0; i < 6; i++ {
		if tallies[i] == 3 {
			_3 = true
			_3_at = i + 1
		}
	}
	if _2 && _3 {
		return _2_at*2 + _3_at*3
	} else {
		return 0
	}
}
